src/pudl/analysis/plant_parts_eia.py

```diff
@@ -1334,56 +1034,83 @@ def add_record_count_per_plant(self, part_df: pd.DataFrame) -> pd.DataFrame:
 
-    def __init__(self, part_name: Literal[PLANT_PARTS_ORDERED]):
-        """Initialize a true granularity labeler."""
-        self.part_name = part_name
-        self.id_cols = PLANT_PARTS[part_name]["id_cols"]
+    The coordinating function here is :meth``execute``.
+    """
 
-    def execute(self, part_df: pd.DataFrame, true_grans):
-        """Merge the true granularity labels into the plant part df.
+    def execute(self, ppl):
+        """Merge the true granularity labels onto the plant part df.
 
-        Args:
-            part_df:
-            true_grans:
+        This method will add the columns ``true_gran``, ``appro_part_label``, and
+        ``appro_record_id_eia`` to the plant parts list which denote whether
+        each plant-part is a true or false granularity.
 
-        """  # noqa: D417
-        # get only the columns you need for this part and drop duplicates
-        bool_df = true_grans[
-            self.id_cols
-            + IDX_TO_ADD
-            + IDX_OWN_TO_ADD
-            + [
-                f"true_gran_{self.part_name}",
-                f"appro_part_label_{self.part_name}",
-                f"record_id_eia_{self.part_name}",
-            ]
-        ].drop_duplicates()
-
-        prop_true_len1 = len(bool_df[bool_df[f"true_gran_{self.part_name}"]]) / len(
-            bool_df
-        )
-        logger.debug(f"proportion of trues: {prop_true_len1:.02}")
-        logger.debug(f"number of records pre-merge:  {len(part_df)}")
+        First the plant part list records are matched to generators. Then
+        the matched records are sorted by PLANT_PARTS_ORDERED and the
+        highest granularity record for each generator is marked as the true
+        granularity. The appropriate true granular part label and record id
+        is then merged on to get the plant part table with true granularity labels.
 
-        part_df = part_df.merge(
-            bool_df,
-            on=self.id_cols + IDX_TO_ADD + IDX_OWN_TO_ADD,
-            how="left",
-        ).rename(
+        Arguments:
+            ppl: (pd.DataFrame) The plant parts list
+        """
+        parts_to_gens = match_to_single_plant_part(
+            multi_gran_df=ppl,
+            ppl=ppl,
+            part_name="plant_gen",
+            cols_to_keep=["plant_part"],
+        )[["record_id_eia_og", "record_id_eia", "plant_part_og"]].rename(
             columns={
-                f"true_gran_{self.part_name}": "true_gran",
-                f"appro_part_label_{self.part_name}": "appro_part_label",
-                f"record_id_eia_{self.part_name}": "appro_record_id_eia",
+                "record_id_eia": "gen_id",
+                "record_id_eia_og": "record_id_eia",
+                "plant_part_og": "plant_part",
             }
         )
+        # concatenate the gen id's to get the combo of gens for each record
+        combos = (
+            parts_to_gens.groupby(["record_id_eia"])["gen_id"]
```

Comment:
should we first `sort_values(["record_id_eia", "gen_id"])`? just in case the `gen_id`'s are mixed up differently within different `record_id_eia`'s? They probably aren't bc of the way `match_to_single_plant_part` is structured, but I would add a sort just in case.

Reply:
Good catch. It should be enough to just sort by `gen_id`? Since the `groupby` on `record_id_eia` preserves ordering within groups. I think it's a little faster to just sort by `gen_id`? I changed this to first sort by `gen_id`.

Reply:
yes that should be sufficient! grand